L6_Pipes/Server/NamedPipe.cpp

```diff
@@ -0,0 +1,80 @@
+#include "NamedPipe.h"
+#include <iostream>
+
+NamedPipe::NamedPipe(const std::string& name, int size)
+{
+	hPipe = CreateNamedPipe(name.c_str(),
+		PIPE_ACCESS_DUPLEX,
+		PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_WAIT,
+		PIPE_UNLIMITED_INSTANCES,
+		size,
+		size,
+		10000,
+		nullptr);
+
+    if (hPipe == INVALID_HANDLE_VALUE)
+    {
+		throw std::exception("Unable to create named pipe");
+    }
+}
+
+NamedPipe NamedPipe::OpenPipe(const std::string& name)
+{
+	NamedPipe pipe;
+
+    pipe.hPipe = CreateFile(name.c_str(),
+		GENERIC_READ | GENERIC_WRITE,
+		0,
+		nullptr,
+		OPEN_EXISTING,
+		0,
+		nullptr);
+
+	if (pipe.hPipe == INVALID_HANDLE_VALUE)
+	{
+		throw std::exception("Unable to open pipe");
+	}
+	pipe.owner = false;
+	return pipe;
+}
+
+NamedPipe::NamedPipe(NamedPipe&& other) noexcept: hPipe(other.hPipe),
+                                                  connected(other.connected),
+                                                  owner(other.owner)
+{
+	other.owner = false;
+}
+
+NamedPipe& NamedPipe::operator=(NamedPipe&& other) noexcept
+{
+    using std::swap;
+	swap(other.hPipe, hPipe);
```

Comment:
do not try to do micro optimisations, it's field of compilator's work

Reply:
It is just a Resharper's suggestion